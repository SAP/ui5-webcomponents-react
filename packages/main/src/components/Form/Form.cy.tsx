import { createPortal } from 'react-dom';
import { InputType } from '../../enums/InputType';
import { Input, Label } from '../../webComponents';
import { FormGroup } from '../FormGroup';
import { FormItem } from '../FormItem';
import { Form } from './index';
import { cypressPassThroughTestsFactory } from '@/cypress/support/utils';

const component = (
  <Form titleText={'Test form'}>
    {false}
    {null}
    {undefined}
    <FormGroup titleText={'Group 1'}>
      <FormItem label={'item 1'}>
        <Input data-testid="formInput" type={InputType.Text} />
      </FormItem>
      <FormItem label={'item 2'}>
        <Input type={InputType.Number} />
      </FormItem>
    </FormGroup>
    <FormGroup titleText={'Group 2'}>
      <FormItem label={'item 3'}>
        <Input data-testid="formInput2" type={InputType.Text} />
      </FormItem>
      <FormItem label={<Label>item 4</Label>}>
        <Input type={InputType.Number} />
      </FormItem>
    </FormGroup>
  </Form>
);

describe('Form', () => {
  it('size S - labels and fields should cover full width', () => {
    cy.viewport(393, 852); // iPhone 14 Pro
    cy.mount(component);
    cy.findByText('item 1:').should('have.css', 'grid-column', '1 / span 12');
    cy.findByTestId('formInput').parent().should('have.css', 'grid-column', 'auto / span 12');
  });

  it('size M - label should cover 2/12, field 10/12', () => {
    cy.viewport(834, 1194); // iPad Pro
    cy.mount(component);
    cy.findByText('item 1:').should('have.css', 'grid-column', '1 / span 2');
    cy.findByTestId('formInput').parent().should('have.css', 'grid-column', 'auto / span 10');
  });

  it('size L - label should cover 1/3, field 2/3', () => {
    cy.viewport(1280, 1024);
    cy.mount(component);
    cy.findByText('item 1:').should('have.css', 'grid-column', '1 / span 4');
    cy.findByTestId('formInput').parent().should('have.css', 'grid-column', 'auto / span 8');
  });

  it('size XL - render two columns with 1/3 and 2/3 each', () => {
    cy.viewport(1920, 1080);
    cy.mount(component);
    cy.findByText('item 1:').should('have.css', 'grid-column', '1 / span 4');
    cy.findByText('item 3:').should('have.css', 'grid-column', '13 / span 4');
    cy.findByTestId('formInput').parent().should('have.css', 'grid-column', '5 / span 8');
    cy.findByTestId('formInput2').parent().should('have.css', 'grid-column', '17 / span 8');
  });

  it('change tag name', () => {
    cy.mount(
      <Form as={'div'} data-testid={'form'}>
        <FormItem>
          <Input type={InputType.Text} />
        </FormItem>
      </Form>
    );
    cy.findByTestId('form').should('have.prop', 'nodeName', 'DIV');
  });

  it('a11y labels', () => {
    cy.mount(component);
    for (let i = 1; i <= 3; i++) {
      cy.findByText(`item ${i}`).should('exist').should('not.be.visible');
      cy.findByText(`item ${i}:`).should('be.visible');
    }
    // custom `Label`
    cy.findAllByText(`item 4`).eq(0).should('be.visible');
    cy.findAllByText(`item 4`).eq(1).should('exist').should('not.be.visible');
  });

  it('FilterItem: doesnt crash with portal as child', () => {
    cy.mount(
      <Form as={'div'} data-testid={'form'}>
        <FormItem label="Portal">
          {createPortal(<Input data-testid="notSupported" type={InputType.Text} />, document.body)}
        </FormItem>
      </Form>
    );
    cy.findByText('Portal:').should('be.visible');
    cy.findByTestId('notSupported').should('not.exist');
  });

  cypressPassThroughTestsFactory(Form, {
    children: (
      <FormItem label="Item">
        <Input type={InputType.Text} />
      </FormItem>
    )
  });
});
