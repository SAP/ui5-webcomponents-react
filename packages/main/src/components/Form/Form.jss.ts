import { ThemingParameters } from '@ui5/webcomponents-react-base';

const labelSpanClasses = () => {
  let styleClasses = {};
  for (let i = 1; i <= 11; i++) {
    styleClasses[`labelSpan${i}`] = {
      '--ui5wcr_form_content_span': 12 - i,
      '--ui5wcr_form_label_span': i
    };
  }
  return styleClasses;
};

const styles = {
  form: {
    display: 'grid',
    alignItems: 'center',
    rowGap: '0.25rem',
    columnGap: '0.5rem',
    gridTemplateColumns: `repeat(12, 1fr)`,
    '--ui5wcr_form_full_span': 'span 12',
    '--ui5wcr_form_label_text_align': 'end',
    '&[data-columns="1"]': {},
    '&[data-columns="2"]': {
      gridTemplateColumns: `repeat(24, 1fr)`,
      '--ui5wcr_form_full_span': 'span 24'
    },
    '&[data-columns="3"]': {
      gridTemplateColumns: `repeat(36, 1fr)`,
      '--ui5wcr_form_full_span': 'span 36'
    },
    '&[data-columns="4"]': {
      gridTemplateColumns: `repeat(48, 1fr)`,
      '--ui5wcr_form_full_span': 'span 48'
    },
    '&[data-columns="5"]': {
      gridTemplateColumns: `repeat(60, 1fr)`,
      '--ui5wcr_form_full_span': 'span 60'
    },
    '&[data-columns="6"]': {
      gridTemplateColumns: `repeat(72, 1fr)`,
      '--ui5wcr_form_full_span': 'span 72'
    },
    '&[data-columns="7"]': {
      gridTemplateColumns: `repeat(84, 1fr)`,
      '--ui5wcr_form_full_span': 'span 84'
    },
    '&[data-columns="8"]': {
      gridTemplateColumns: `repeat(96, 1fr)`,
      '--ui5wcr_form_full_span': 'span 96'
    },
    '&[data-columns="9"]': {
      gridTemplateColumns: `repeat(108, 1fr)`,
      '--ui5wcr_form_full_span': 'span 108'
    },
    '&[data-columns="10"]': {
      gridTemplateColumns: `repeat(120, 1fr)`,
      '--ui5wcr_form_full_span': 'span 120'
    },
    '&[data-columns="11"]': {
      gridTemplateColumns: `repeat(132, 1fr)`,
      '--ui5wcr_form_full_span': 'span 132'
    },
    '&[data-columns="12"]': {
      gridTemplateColumns: `repeat(144, 1fr)`,
      '--ui5wcr_form_full_span': 'span 144'
    }
  },
  formTitle: {
    borderBottom: `1px solid ${ThemingParameters.sapGroup_TitleBorderColor}`,
    marginBottom: '1.75rem',
    gridColumn: 'var(--ui5wcr_form_full_span)'
  },
  ...labelSpanClasses(),
  labelSpan12: {
    '--ui5wcr_form_content_span': 12,
    '--ui5wcr_form_label_text_align': 'start',
    '--ui5wcr_form_label_span': 12,
    rowGap: 0
  }
};

export { styles };
