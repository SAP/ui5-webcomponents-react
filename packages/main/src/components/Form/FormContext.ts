import { createContext, useContext } from 'react';

export const FormContext = createContext({ labelSpan: null });

export function useFormContext() {
  return useContext(FormContext);
}
