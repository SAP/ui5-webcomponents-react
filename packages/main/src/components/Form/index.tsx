import { CssSizeVariables, Device, ThemingParameters, useSyncRef } from '@ui5/webcomponents-react-base';
import { clsx } from 'clsx';
import React, {
  Children,
  cloneElement,
  CSSProperties,
  ElementType,
  forwardRef,
  ReactElement,
  ReactNode,
  useEffect,
  useMemo,
  useRef,
  useState
} from 'react';
import { createUseStyles } from 'react-jss';
import { FormBackgroundDesign, TitleLevel } from '../../enums';
import { CommonProps } from '../../interfaces/CommonProps';
import { Title } from '../../webComponents/Title';
import { styles } from './Form.jss';

export interface FormPropTypes extends CommonProps {
  /**
   * Components that are placed into Form.
   *
   * __Note:__ Although this prop accepts all HTML Elements, it is strongly recommended that you only use `FormGroup` or `FormItem` in order to preserve the intended design.
   */
  children: ReactNode | ReactNode[];
  /**
   * Specifies the background color of the Form content.
   */
  backgroundDesign?: FormBackgroundDesign;
  /**
   * Form title
   */
  titleText?: string;
  /**
   * Form columns for small size (`< 600px`).
   * Must be a number between 1 and 12.
   *
   * Default Value: 1
   */
  columnsS?: number;
  /**
   * Form columns for medium size (`600px` - `1023px`).
   * Must be a number between 1 and 12.
   *
   * Default Value: 1
   *
   * __Note__: The number of columns for medium size must not be smaller than the number of columns for small size.
   */
  columnsM?: number;
  /**
   * Form columns for large size (`1024px` - `1439px`).
   * Must be a number between 1 and 12.
   *
   * Default Value: 1
   *
   * __Note:__ The number of columns for large size must not be smaller than the number of columns for medium size.
   */
  columnsL?: number;
  /**
   * Form columns for extra large size (`>= 1440px`).
   * Must be a number between 1 and 12.
   *
   * Default Value: 2
   *
   * __Note:__ The number of columns for extra large size must not be smaller than the number of columns for large size.
   */
  columnsXL?: number;

  /**
   * Default span for labels in small size (`< 600px`).
   * Must be a number between 1 and 12.
   *
   * Default Value: 12
   */
  labelSpanS?: number;
  /**
   * Default span for labels in medium size (`600px` - `1023px`).
   * Must be a number between 1 and 12.
   *
   * Default Value: 2
   */
  labelSpanM?: number;
  /**
   * Default span for labels in large size (`1024px` - `1439px`).
   * Must be a number between 1 and 12.
   *
   * Default Value: 4
   */
  labelSpanL?: number;
  /**
   * Default span for labels in extra large size (`>= 1440px`).
   * Must be a number between 1 and 12.
   *
   * Default Value: 4
   */
  labelSpanXL?: number;
  /**
   * Sets the components outer HTML tag.
   *
   * __Note:__ For TypeScript the types of `ref` are bound to the default tag name, if you change it you are responsible to set the respective types yourself.
   */
  as?: keyof HTMLElementTagNameMap;
}

const clonedChildrenForSingleColumn = (reactChildren, currentLabelSpan) =>
  Children.map(reactChildren, (child) => {
    if (child?.type?.displayName === 'FormItem') {
      return cloneElement(child, { labelSpan: currentLabelSpan });
    }
    if (child?.type?.displayName === 'FormGroup') {
      return cloneElement(child, { children: clonedChildrenForSingleColumn(child.props.children, currentLabelSpan) });
    }
    return child;
  });

const useStyles = createUseStyles(styles, { name: 'Form' });
/**
 * The `Form` component arranges labels and fields into groups and rows. There are different ways to visualize forms for different screen sizes.
 * It is possible to change the alignment of all labels by setting the CSS `align-items` property, per default all labels are centered.
 */
const Form = forwardRef<HTMLFormElement, FormPropTypes>((props, ref) => {
  const {
    as,
    backgroundDesign,
    children,
    columnsS,
    columnsM,
    columnsL,
    columnsXL,
    className,
    labelSpanS,
    labelSpanM,
    labelSpanL,
    labelSpanXL,
    slot,
    titleText,
    style,
    ...rest
  } = props;

  const columnsMap = new Map();
  columnsMap.set('Phone', columnsS);
  columnsMap.set('Tablet', columnsM);
  columnsMap.set('Desktop', columnsL);
  columnsMap.set('LargeDesktop', columnsXL);

  const labelSpanMap = new Map();
  labelSpanMap.set('Phone', labelSpanS);
  labelSpanMap.set('Tablet', labelSpanM);
  labelSpanMap.set('Desktop', labelSpanL);
  labelSpanMap.set('LargeDesktop', labelSpanXL);

  const [componentRef, formRef] = useSyncRef<HTMLFormElement>(ref);
  // use the window range set as first best guess, if not available use Desktop
  const [currentRange, setCurrentRange] = useState(Device.getCurrentRange()?.name ?? 'Desktop');
  const lastRange = useRef(currentRange);

  useEffect(() => {
    const observer = new ResizeObserver(([form]) => {
      const rangeInfo = Device.getCurrentRange(form.contentRect.width);
      if (rangeInfo && lastRange.current !== rangeInfo.name) {
        lastRange.current = rangeInfo.name;
        setCurrentRange(rangeInfo.name);
      }
    });

    if (formRef.current) {
      observer.observe(formRef.current);
    }

    return () => {
      observer.disconnect();
    };
  }, [formRef, setCurrentRange, lastRange]);

  const classes = useStyles();

  const currentNumberOfColumns = columnsMap.get(currentRange);
  const currentLabelSpan = labelSpanMap.get(currentRange);

  const [formGroups, updatedTitle] = useMemo(() => {
    const computedFormGroups = [];
    if (Children.count(children) === 1 && !titleText) {
      const singleChild = (Array.isArray(children) ? children[0] : children) as ReactElement;
      if (singleChild?.props?.title?.length > 0) {
        return [cloneElement(singleChild, { title: null }), singleChild.props.title];
      }
    }

    const currentColumnCount = currentNumberOfColumns;
    if (currentColumnCount === 1) {
      return [clonedChildrenForSingleColumn(children, currentLabelSpan), titleText];
    }

    const rows = [];
    const childrenArray = Children.toArray(children);
    const estimatedNumberOfGroupRows = childrenArray.length / currentColumnCount;
    for (let i = 0; i < estimatedNumberOfGroupRows; i++) {
      rows[i] = childrenArray.slice(i * currentColumnCount, i * currentColumnCount + currentColumnCount);
    }

    const maxRowsPerRow: number[] = [];
    rows.forEach((rowGroup: ReactElement[], rowIndex) => {
      const numberOfRowsOfEachForm = rowGroup.map((row) => {
        if ((row.type as any).displayName === 'FormItem') {
          return 1;
        }
        return Children.count(row.props.children) + 1;
      });

      maxRowsPerRow[rowIndex] = Math.max(...numberOfRowsOfEachForm);
    });

    let totalRowCount = 2;

    rows.forEach((column: ReactElement[], rowIndex) => {
      const rowsForThisRow = maxRowsPerRow[rowIndex];
      column.forEach((cell, columnIndex) => {
        const titleStyles: CSSProperties = {
          gridColumnEnd: 'span 12',
          gridColumnStart: columnIndex * 12 + 1,
          gridRowStart: totalRowCount,
          display: 'flex',
          alignItems: 'center',
          fontFamily: ThemingParameters.sapFontFamily,
          height: CssSizeVariables.sapWcrFormGroupTitleHeight,
          lineHeight: CssSizeVariables.sapWcrFormGroupTitleHeight,
          color: ThemingParameters.sapTextColor,
          fontSize: ThemingParameters.sapFontSize,
          fontWeight: 'bold',
          backgroundColor: ThemingParameters.sapGroup_TitleBackground,
          margin: 0,
          paddingTop: '1rem'
        };

        if (cell?.props?.titleText) {
          computedFormGroups.push(
            <h6
              style={titleStyles}
              title={cell.props.titleText}
              aria-label={cell.props.titleText}
              key={`title-col-${columnIndex}-row-${totalRowCount}`}
            >
              {cell.props.titleText}
            </h6>
          );
        }

        for (let i = 0; i < rowsForThisRow; i++) {
          const itemToRender =
            (cell.type as any).displayName === 'FormGroup'
              ? Children.toArray(cell.props.children)[i]
              : (cell.type as any).displayName === 'FormItem' && i === 0
              ? cell
              : null;

          if (itemToRender) {
            computedFormGroups.push(
              cloneElement(itemToRender as ReactElement, {
                key: `col-${columnIndex}-row-${totalRowCount + i}`,
                columnIndex,
                lastGroupItem: (cell.type as any).displayName === 'FormGroup' && rowsForThisRow - 2 === i,
                rowIndex: totalRowCount + i + 1,
                labelSpan: currentLabelSpan
              })
            );
          }
        }
      });
      totalRowCount += rowsForThisRow;
      if (rowsForThisRow === 1) {
        totalRowCount += 1;
      }
    });

    return [computedFormGroups, titleText];
  }, [children, currentRange, titleText, currentNumberOfColumns, currentLabelSpan]);

  const formClassNames = clsx(
    classes.form,
    classes[`labelSpan${((currentLabelSpan - 1) % 12) + 1}`],
    classes[backgroundDesign.toLowerCase()],
    className
  );

  const CustomTag = as as ElementType;
  return (
    <CustomTag
      ref={componentRef}
      slot={slot}
      className={formClassNames}
      style={style}
      data-columns={currentNumberOfColumns}
      {...rest}
    >
      {updatedTitle && (
        <Title level={TitleLevel.H3} className={classes.formTitle}>
          {updatedTitle}
        </Title>
      )}
      {formGroups}
    </CustomTag>
  );
});

Form.displayName = 'Form';

Form.defaultProps = {
  as: 'form',
  backgroundDesign: FormBackgroundDesign.Transparent,
  columnsS: 1,
  columnsM: 1,
  columnsL: 1,
  columnsXL: 2,
  labelSpanS: 12,
  labelSpanM: 2,
  labelSpanL: 4,
  labelSpanXL: 4
};

export { Form };
