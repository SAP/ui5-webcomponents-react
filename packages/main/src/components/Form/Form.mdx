import { ControlsWithNote, DocsHeader, Footer } from '@sb/components';
import SubcomponentsSection from '@sb/docs/SubcomponentsSection.md?raw';
import { ArgTypes, Canvas, Description, Markdown, Meta } from '@storybook/blocks';
import { FormGroup } from '../FormGroup';
import { FormItem } from '../FormItem';
import * as ComponentStories from './Form.stories';

<Meta of={ComponentStories} />

<DocsHeader since="0.7.0" />

<br />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

<Markdown>{SubcomponentsSection}</Markdown>

## Form Group

<Description of={FormGroup} />

<ArgTypes of={FormGroup} />

## FormItem

<Description of={FormItem} />

<ArgTypes of={FormItem} />

<Footer />
