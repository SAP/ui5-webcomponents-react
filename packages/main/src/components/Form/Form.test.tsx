import { render } from '@shared/tests';
import { createPassThroughPropsTest } from '@shared/tests/utils';
import { Form } from '@ui5/webcomponents-react/dist/Form';
import { FormGroup } from '@ui5/webcomponents-react/dist/FormGroup';
import { FormItem } from '@ui5/webcomponents-react/dist/FormItem';
import { Input } from '@ui5/webcomponents-react/dist/Input';
import { InputType } from '@ui5/webcomponents-react/dist/InputType';
import { Label } from '@ui5/webcomponents-react/dist/Label';
import * as React from 'react';

const SIZE_S = 200;
const SIZE_M = 800;
const SIZE_L = 1200;
const SIZE_XL = 1600;
const component = (
  <Form titleText={'Test form'}>
    <FormGroup titleText={'Group 1'}>
      <FormItem label={'item 1'}>
        <Input type={InputType.Text} />
      </FormItem>
      <FormItem label={'item 2'}>
        <Input type={InputType.Number} />
      </FormItem>
    </FormGroup>
    <FormGroup titleText={'Group 2'}>
      <FormItem label={'item 1'}>
        <Input type={InputType.Text} />
      </FormItem>
      <FormItem label={'item 2'}>
        <Input type={InputType.Number} />
      </FormItem>
    </FormGroup>
  </Form>
);

describe('Create a Form', () => {
  test('size rate S; should create Label and Element with 100% width and display: block for top FormItem div', () => {
    window = Object.assign(window, { innerWidth: SIZE_S });
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });

  test('size rate M; should create Label and Element with 16% and 83% width respectively and display: flex for top FormItem div', () => {
    window = Object.assign(window, { innerWidth: SIZE_M });
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });

  test('size rate L; should create Label and Element with 33% and 66% width respectively and display: flex for top FormItem div', () => {
    window = Object.assign(window, { innerWidth: SIZE_L });
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });

  test('size rate XL; should create Label and Element with 33% and 66% width respectively and display: flex for top FormItem div', () => {
    window = Object.assign(window, { innerWidth: SIZE_XL });
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should use a single FormGroup's heading as a Form heading if one is not set", () => {
    const ungroupedChildren = (
      <Form>
        <FormGroup titleText={'To be Form title'}>
          <FormItem label={'item 1'}>
            <Input type={InputType.Text} />
          </FormItem>
          <FormItem label={'item 2'}>
            <Input type={InputType.Number} />
          </FormItem>
        </FormGroup>
      </Form>
    );
    const { asFragment } = render(ungroupedChildren);
    expect(asFragment()).toMatchSnapshot();
  });

  createPassThroughPropsTest(Form, {
    children: (
      <FormItem>
        <Input type={InputType.Text} />
      </FormItem>
    )
  });
});
