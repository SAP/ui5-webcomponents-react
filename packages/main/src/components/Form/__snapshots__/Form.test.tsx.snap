// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create a Form should use a single FormGroup's heading as a Form heading if one is not set 1`] = `
<DocumentFragment>
  <form
    class="Form-form Form-labelSpan4"
    data-columns="2"
  >
    <h6
      aria-label="To be Form title"
      style="grid-column-end: span 12; grid-column-start: 1; grid-row-start: 2; display: flex; align-items: center; font-family: var(--sapFontFamily); line-height: var(--sapWcrFormGroupTitleHeight); font-weight: bold; margin: 0px; padding-top: 1rem;"
      title="To be Form title"
    >
      To be Form title
    </h6>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 3; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 1:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-column-start="5"
      data-grid-row-start="3"
      style="grid-column-start: 5; grid-row-start: 3; padding-bottom: 0px;"
    >
      <ui5-input
        type="Text"
        ui5-input=""
        value-state="None"
      />
    </div>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 4; padding-bottom: 1rem;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 2:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-column-start="5"
      data-grid-row-start="4"
      style="grid-column-start: 5; grid-row-start: 4; padding-bottom: 1rem;"
    >
      <ui5-input
        type="Number"
        ui5-input=""
        value-state="None"
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`Create a Form size rate L; should create Label and Element with 33% and 66% width respectively and display: flex for top FormItem div 1`] = `
<DocumentFragment>
  <form
    class="Form-form Form-labelSpan4"
    data-columns="1"
  >
    <ui5-title
      class="Form-formTitle"
      level="H3"
      ui5-title=""
    >
      Test form
    </ui5-title>
    <h6
      aria-label="Group 1"
      class="FormGroup-title"
      title="Group 1"
    >
      Group 1
    </h6>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 1:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Text"
        ui5-input=""
        value-state="None"
      />
    </div>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 2:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Number"
        ui5-input=""
        value-state="None"
      />
    </div>
    <span
      class="FormGroup-spacer"
    />
    <h6
      aria-label="Group 2"
      class="FormGroup-title"
      title="Group 2"
    >
      Group 2
    </h6>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 1:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Text"
        ui5-input=""
        value-state="None"
      />
    </div>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 2:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Number"
        ui5-input=""
        value-state="None"
      />
    </div>
    <span
      class="FormGroup-spacer"
    />
  </form>
</DocumentFragment>
`;

exports[`Create a Form size rate M; should create Label and Element with 16% and 83% width respectively and display: flex for top FormItem div 1`] = `
<DocumentFragment>
  <form
    class="Form-form Form-labelSpan2"
    data-columns="1"
  >
    <ui5-title
      class="Form-formTitle"
      level="H3"
      ui5-title=""
    >
      Test form
    </ui5-title>
    <h6
      aria-label="Group 1"
      class="FormGroup-title"
      title="Group 1"
    >
      Group 1
    </h6>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 1:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Text"
        ui5-input=""
        value-state="None"
      />
    </div>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 2:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Number"
        ui5-input=""
        value-state="None"
      />
    </div>
    <span
      class="FormGroup-spacer"
    />
    <h6
      aria-label="Group 2"
      class="FormGroup-title"
      title="Group 2"
    >
      Group 2
    </h6>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 1:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Text"
        ui5-input=""
        value-state="None"
      />
    </div>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 2:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Number"
        ui5-input=""
        value-state="None"
      />
    </div>
    <span
      class="FormGroup-spacer"
    />
  </form>
</DocumentFragment>
`;

exports[`Create a Form size rate S; should create Label and Element with 100% width and display: block for top FormItem div 1`] = `
<DocumentFragment>
  <form
    class="Form-form Form-labelSpan12"
    data-columns="1"
  >
    <ui5-title
      class="Form-formTitle"
      level="H3"
      ui5-title=""
    >
      Test form
    </ui5-title>
    <h6
      aria-label="Group 1"
      class="FormGroup-title"
      title="Group 1"
    >
      Group 1
    </h6>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 1:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Text"
        ui5-input=""
        value-state="None"
      />
    </div>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 2:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Number"
        ui5-input=""
        value-state="None"
      />
    </div>
    <span
      class="FormGroup-spacer"
    />
    <h6
      aria-label="Group 2"
      class="FormGroup-title"
      title="Group 2"
    >
      Group 2
    </h6>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 1:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Text"
        ui5-input=""
        value-state="None"
      />
    </div>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 0; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 2:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-row-start="0"
      style="grid-row-start: 0; padding-bottom: 0px;"
    >
      <ui5-input
        type="Number"
        ui5-input=""
        value-state="None"
      />
    </div>
    <span
      class="FormGroup-spacer"
    />
  </form>
</DocumentFragment>
`;

exports[`Create a Form size rate XL; should create Label and Element with 33% and 66% width respectively and display: flex for top FormItem div 1`] = `
<DocumentFragment>
  <form
    class="Form-form Form-labelSpan4"
    data-columns="2"
  >
    <ui5-title
      class="Form-formTitle"
      level="H3"
      ui5-title=""
    >
      Test form
    </ui5-title>
    <h6
      aria-label="Group 1"
      style="grid-column-end: span 12; grid-column-start: 1; grid-row-start: 2; display: flex; align-items: center; font-family: var(--sapFontFamily); line-height: var(--sapWcrFormGroupTitleHeight); font-weight: bold; margin: 0px; padding-top: 1rem;"
      title="Group 1"
    >
      Group 1
    </h6>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 3; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 1:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-column-start="5"
      data-grid-row-start="3"
      style="grid-column-start: 5; grid-row-start: 3; padding-bottom: 0px;"
    >
      <ui5-input
        type="Text"
        ui5-input=""
        value-state="None"
      />
    </div>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 1; grid-row-start: 4; padding-bottom: 1rem;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 2:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-column-start="5"
      data-grid-row-start="4"
      style="grid-column-start: 5; grid-row-start: 4; padding-bottom: 1rem;"
    >
      <ui5-input
        type="Number"
        ui5-input=""
        value-state="None"
      />
    </div>
    <h6
      aria-label="Group 2"
      style="grid-column-end: span 12; grid-column-start: 13; grid-row-start: 2; display: flex; align-items: center; font-family: var(--sapFontFamily); line-height: var(--sapWcrFormGroupTitleHeight); font-weight: bold; margin: 0px; padding-top: 1rem;"
      title="Group 2"
    >
      Group 2
    </h6>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 13; grid-row-start: 3; padding-bottom: 0px;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 1:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-column-start="17"
      data-grid-row-start="3"
      style="grid-column-start: 17; grid-row-start: 3; padding-bottom: 0px;"
    >
      <ui5-input
        type="Text"
        ui5-input=""
        value-state="None"
      />
    </div>
    <ui5-label
      class="FormItem-label"
      style="grid-column-start: 13; grid-row-start: 4; padding-bottom: 1rem;"
      ui5-label=""
      wrapping-type="Normal"
    >
      item 2:
    </ui5-label>
    <div
      class="FormItem-content"
      data-grid-column-start="17"
      data-grid-row-start="4"
      style="grid-column-start: 17; grid-row-start: 4; padding-bottom: 1rem;"
    >
      <ui5-input
        type="Number"
        ui5-input=""
        value-state="None"
      />
    </div>
  </form>
</DocumentFragment>
`;
